thoughts-of-a-rustacean-learning-go.md

```diff
@@ -0,0 +1,244 @@
+and whilst it's not very different from `.send()` and `.recv()` in Rust, I found it surprisingly easy to read.
+Initially I got confused often by which side the channel was, but after a while I got used to it.
+It also has an in built select block for selecting over channels (Rust has a macro for the same purpose, but no language support).
+
+
+`gofmt`. The Go style of coding is different from the Rust one (tabs vs spaces, how declarations look),
+but I continued to use the Rust style because of the muscle memory (also too lazy to change the settings
+in my editor). `gofmt` made life easy since I could just run it in a directory and it would fix everything.
+Eventually I was able to learn the proper style by watching my code get corrected.
+I'd love to see a `rustfmt`, however!
+
+
+Go is great for debugging programs with multiple threads, too. It can detect deadlocks and
+post traces for the threads (with metadata including what code the thread was spawned from, as well as its
+current state). It also posts such traces when the program crashes. These are totally awesome and saved me tons of time whilst debugging
+my code (which at times had all sorts of cross interactions between more than ten goroutines in the tests)
+Without a green threading framework, I'm not sure how easy it will be to integrate this into Rust
+(for debug builds, obviously), but I'd certainly like it to be there some day.
+
+Go has really great <s>green threads</s> goroutines. They're rather efficient
+(I can spawn a thousand and it schedules them nicely), and easy to use.
+
+Go has really good in built support and tooling for tests (Rust does too).
+I enjoyed writing tests in Go quite a bit due to this.
+
+## What I didn't like
+
+Sadly, there are a lot of things here, but bear in mind what I mentioned above
+about me being new to Go and not yet familiar with the "Go way" of doing things.
+
+### No enums
+
+Rust has enums, which are basically tagged unions. Different variants can contain different types
+of data, so we can have, for example:
+
+```rust
+enum Shape {
+    Rectangle(Point, Point),
+    Circle(Point, u8),
+    Triangle(Point, Point, Point)
+}
+```
+
+and when matching/destructuring, you get type-safe access to the contents of the variant.
+
+This is extremely useful for sending typed messages across channels. For example, in Servo we use such
+[an enum for sending details about the progress of a fetch to the corresponding XHR object][servo-xhr-enum].
+[Another such enum][servo-constellation-enum] is used for communication between the constellation and the compositor/script.
+
+
+This gives us a great degree of type safety; I can send messages with different data within them,
+however I can only send messages that the other end will know how to handle since they must all be of the type of the message enum.
+
+In Go there's no obvious way to get this. The closest thing is the type called `interface {}`
```

Comment:
I'd again be interested in seeing some larger side-by-side code comparisons here. It would help me get a better idea of how you would structure data for passing messages between threads in Rust vs. Go-routines in Go.